File: recursion/computing-powers/powers.py
def isEven(n):
    return n % 2 == 0


def isOdd(n):
    return not isEven(n)


def power(x, n):
    """Computes x to the power of n

    Args:
        x (number):  Base
        n (int): Exponent
    Returns:
        (number) x^n
    """
    # base case
    if n == 0:
        return 1

    # n = negative
    # MUST COME BEFORE ODD/EVEN CHECKS
    # isOdd/Even will return True for negatives
    # Exponent wont get smaller -> won't terminate
    if(n < 0):
        return 1 / power(x, -n)
    # n = positive and odd
    if(isOdd(n)):
        return x * power(x, n - 1)

    # n = positive and even
    if (isEven(n)):
        y = power(x, n / 2)
        return y * y


def display_power(x, n):
    print("%s^%s = %s" % (x, n, power(x, n)))


display_power(3, 0)
assert(power(3, 0) == 1)

display_power(-3, 0)
assert(power(-3, 0) == 1)

display_power(3, 1)
assert(power(3, 1) == 3)

display_power(3, 2)
assert(power(3, 2) == 9)

display_power(3, 3)
assert(power(3, 3) == 27)

display_power(3, -2)
assert(power(3, -2) == 1 / 9)


display_power(-3, -2)
assert(power(-3, -2) == 1 / 9)

File: recursion/computing-powers/powers.js
function isEven(n) {
    return n % 2 === 0;
};

function isOdd(n) {
    return !isEven(n);
};

function power(x, n) {
    console.log("Computing " + x + " raised to power " + n + ".");
    // base case
    if (n === 0) {
        return 1;
    }
    // recursive case: n is negative
    if (n < 0) {
        return 1 / power(x, -n);
    }
    // recursive case: n is odd
    if (isOdd(n)) {
        return x * power(x, n - 1);
    }
    // recursive case: n is even
    if (isEven(n)) {
        let y = power(x, n / 2);
        return y * y;
    }
};

function displayPower(x, n) {
    console.log(x + " to the " + n + " is " + power(x, n));
};

displayPower(3, 0);
console.assert(power(3, 0) === 1);
displayPower(3, 1);
console.assert(power(3, 1) === 3);
displayPower(3, 2);
console.assert(power(3, 2) === 9);
displayPower(3, -1);
console.assert(power(3, -1) === 1 / 3);
displayPower(2, 6);
console.assert(power(2, 6) === 64);
displayPower(2, 6);
console.assert(power(2, -6) === 1 / 64);